EventFilter/TotemRawToDigi/plugins/TotemVFATRawToDigi.cc

```diff
@@ -0,0 +1,160 @@
+#include "DataFormats/TotemDigi/interface/TotemFEDInfo.h"
+
+#include "CondFormats/DataRecord/interface/TotemReadoutRcd.h"
+#include "CondFormats/TotemReadoutObjects/interface/TotemDAQMapping.h"
+#include "CondFormats/TotemReadoutObjects/interface/TotemAnalysisMask.h"
+
+#include "EventFilter/TotemRawToDigi/interface/SimpleVFATFrameCollection.h"
+#include "EventFilter/TotemRawToDigi/interface/RawDataUnpacker.h"
+#include "EventFilter/TotemRawToDigi/interface/RawToDigiConverter.h"
+
+#include <string>
+
+//----------------------------------------------------------------------------------------------------
+
+class TotemVFATRawToDigi : public edm::one::EDProducer<>
+{
+  public:
+    explicit TotemVFATRawToDigi(const edm::ParameterSet&);
+    ~TotemVFATRawToDigi();
+
+    virtual void produce(edm::Event&, const edm::EventSetup&) override;
+    virtual void endJob();
+
+  private:
+    std::string subSystem;
+
+    std::vector<unsigned int> fedIds;
+
+    edm::EDGetTokenT<FEDRawDataCollection> fedDataToken;
+
+    RawDataUnpacker rawDataUnpacker;
+    RawToDigiConverter rawToDigiConverter;
+
+    template <typename DigiType>
+    void run(edm::Event&, const edm::EventSetup&);
+};
+
+//----------------------------------------------------------------------------------------------------
+
+using namespace edm;
+using namespace std;
+
+//----------------------------------------------------------------------------------------------------
+
+TotemVFATRawToDigi::TotemVFATRawToDigi(const edm::ParameterSet &conf):
+  subSystem(conf.getParameter<string>("subSystem")),
+  fedIds(conf.getParameter< vector<unsigned int> >("fedIds")),
+  rawDataUnpacker(conf.getParameterSet("RawUnpacking")),
+  rawToDigiConverter(conf.getParameterSet("RawToDigi"))
+{
+  fedDataToken = consumes<FEDRawDataCollection>(conf.getParameter<edm::InputTag>("rawDataTag"));
+
+  // validate chosen subSystem
+  if (subSystem != "RP")
+    throw cms::Exception("TotemVFATRawToDigi::TotemVFATRawToDigi") << "Unknown sub-system string " << subSystem << "." << endl;
+
+  // FED (OptoRx) headers and footers
+  produces< vector<TotemFEDInfo> >(subSystem);
+
+  // digi
+  if (subSystem == "RP")
+    produces< DetSetVector<TotemRPDigi> >(subSystem);
+
+  // set default IDs
+  if (fedIds.empty())
+  {
+    if (subSystem == "RP")
+    {
+      for (int id = FEDNumbering::MINTotemRPFEDID; id <= FEDNumbering::MAXTotemRPFEDID; ++id)
+        fedIds.push_back(id);
+    }
+  }
+
+  // conversion status
+  produces< DetSetVector<TotemVFATStatus> >(subSystem);
+}
+
+//----------------------------------------------------------------------------------------------------
+
+TotemVFATRawToDigi::~TotemVFATRawToDigi()
+{
+}
+
+//----------------------------------------------------------------------------------------------------
+
+void TotemVFATRawToDigi::produce(edm::Event& event, const edm::EventSetup &es)
+{
+  if (subSystem == "RP")
+    run< DetSetVector<TotemRPDigi> >(event, es);
+}
```

Comment:
``` TotemFEDInfo ``` and ```TotemVFATStatus ``` are declared in produces regardless of the value of subSystem.
In this case they should be filled/put in the event regardless of the value of subSystem